### PR TITLE
Ugur implementation

### PR DESCRIPTION
**Pull request by murerkinn**

### CODE REVIEW

File: frontend/src/App.vue [hunk 1]
```diff
@@ -67,6 +67,18 @@ h1 {
   background-color: #f6f6f6;
 }
 
+// Apply styles only for monitor route
+.monitor {
```
Comment (dashersw) [inline frontend/src/App.vue]: I think these would be better under the Monitor component — we can use global styles there.

File: frontend/src/components/Ugur.vue [hunk 2]
```diff
@@ -0,0 +1,122 @@
+<script>
+import { mapState, mapActions } from 'vuex'
+
+export default {
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: Let's call this component Director — we also need to add the `name` property.

File: frontend/src/components/Ugur.vue [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  data() {
+    return {
+      triggers: [''],
+      lang: 'tr-TR'
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: the default language should be en-US.

File: frontend/src/components/Ugur.vue [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    a-form(:form="form")
+      a-form-item.ugur-form-item(
+        v-for="(trigger, triggerIndex) in triggers"
+        :key="triggerIndex"
+      )
+        a-input(v-decorator="[`names[${triggerIndex}]`,{validateTrigger: ['change', 'blur'],rules:[{message: 'some message'}]}]" @change="(e) => handleTriggerChange(triggerIndex, e.target.value)" style="width: 85%;")
+        a-button.ugur-action.left(@click="() => removeTrigger({index: triggerIndex})")
+          a-icon(type="minus")
+    a-button.ugur-action.first(@click="handleMoveOn") Done, move on
+    a-button.ugur-action(@click="handleVoiceActivation") Activate: Voice-Action
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: can we call this function `activateVoiceRecognition`?

File: frontend/src/components/Ugur.vue [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+  #ugur
+    a-form(:form="form")
+      a-form-item.ugur-form-item(
+        v-for="(trigger, triggerIndex) in triggers"
+        :key="triggerIndex"
+      )
+        a-input(v-decorator="[`names[${triggerIndex}]`,{validateTrigger: ['change', 'blur'],rules:[{message: 'some message'}]}]" @change="(e) => handleTriggerChange(triggerIndex, e.target.value)" style="width: 85%;")
+        a-button.ugur-action.left(@click="() => removeTrigger({index: triggerIndex})")
+          a-icon(type="minus")
+    a-button.ugur-action.first(@click="handleMoveOn") Done, move on
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: we can call this method `skipQuestion`

File: frontend/src/components/Ugur.vue [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+  data() {
+    return {
+      triggers: [''],
+      lang: 'tr-TR'
+    }
+  },
+  computed: {
+    ...mapState('event', ['event'])
+  },
+  props: ['recognition', 'pinLatestQuestion'],
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: Is it possible to remove recognition prop and call window.SpeechRecognition here?

File: frontend/src/components/Ugur.vue [hunk 7]
```diff
@@ -0,0 +1,122 @@
...
+      const { recognition } = this
+
+      if (!recognition) return console.error('Your browser is not supporting speech recognition api')
+
+      recognition.lang = this.lang
+
+      try {
+        recognition.start()
+      } catch (error) {
+        console.warn('Speech recognition API is already started.')
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: Is this the only possible error? I think we'd benefit from logging the original error instead.

File: frontend/src/components/Ugur.vue [hunk 8]
```diff
@@ -0,0 +1,122 @@
...
+      })
+
+      if (getLastPinnedQuestion()) return
+
+      this.pinLatestQuestion()
+    },
+    handleVoiceTrigger(e) {
+      const transcript = Array.from(e.results)
+        .map(result => result[0])
+        .map(result => result.transcript)
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: please combine these two map calls.

File: frontend/src/views/EventDetail.vue [hunk 9]
```diff
@@ -85,6 +101,16 @@ export default {
       question.votes += question.voted ? 1 : -1
       this.sortQuestions()
     },
+    handleArchive(question) {
```
Comment (dashersw) [inline frontend/src/views/EventDetail.vue:81]: I believe this should be another commit.

File: frontend/src/views/EventDetail.vue [hunk 10]
```diff
@@ -150,8 +182,24 @@ export default {
...
+              placeholder="Type your question"
+              :autoSize="{ minRows: 2, maxRows: 6 }"
+              :maxLength="280"
+              v-model="question"
+            )
+            a-input(placeholder="Your name (optional)" v-model="name" :maxLength="40")
+            a-button(type="primary" @click="sendQuestion" :loading="loading" icon="message") Send
+      a-tab-pane(tab="Ugurcum" key="2")
+        a-card
+          Ugur(:recognition="speechRecognitionInstance" :handlePin="pinLatestQuestion")
```
Comment (dashersw) [inline frontend/src/views/EventDetail.vue]: passing `:handlePin` here makes it very difficult to follow — let's implement this as a store action.

File: frontend/src/views/EventDetail.vue [hunk 11]
```diff
@@ -150,8 +182,24 @@ export default {
   .content
     h1 {{ event.title }}
     h3 {{ event.description }}
-    a-card
-      form(@submit.prevent='sendQuestion')
+    a-tabs(v-if="event.owner == user._id")
+      a-tab-pane(tab="Ask the speaker" key="1")
+        a-card
```
Comment (dashersw) [inline frontend/src/views/EventDetail.vue]: this card needs to become a separate component so that you can reuse it.

Comment (dashersw) [reply frontend/src/views/EventDetail.vue]: ...or, the form below :)

File: frontend/src/views/EventDetail.vue (hunk 9 above)
Comment (dashersw) [reply frontend/src/views/EventDetail.vue:81]: Yep, now checked — this comes before the backend implementation, so please separate this :) you can even combine it with the other backend implementation commit.

File: frontend/src/components/Ugur.vue (hunk 3 above)
Comment (dashersw) [reply frontend/src/components/Ugur.vue]: it would be great if we can read this from the user, somehow. adding a property to the user for this is not far-fetched.

File: frontend/src/components/Ugur.vue [hunk 12]
```diff
@@ -0,0 +1,122 @@
+<script>
+import { mapState, mapActions } from 'vuex'
+
+export default {
+  data() {
+    return {
+      triggers: [''],
```
Comment (dashersw) [inline frontend/src/components/Ugur.vue]: this as well should be stored on the user object in the database.

File: backend/src/models/event.js [hunk 13]
```diff
@@ -58,14 +58,18 @@ Event.pre('save', async function (next) {
...
 Event.static('decorateForUser', function (event, userIds) {
-  const updatedQuestions = event.questions.toObject().map(q => {
-    return {
-      ...q,
-      voted: q.voters.some(v => userIds.some(id => id.equals(v))),
-      ownQuestion: userIds.some(i => i.equals(q.user)),
-      user: undefined,
-      voters: undefined,
-    }
+  const updatedQuestions = event
```
Comment (dashersw) [inline backend/src/models/event.js:63]: this commit should be in archive backend implementation.

File: backend/src/models/event.js [hunk 14]
```diff
@@ -58,14 +58,18 @@ Event.pre('save', async function (next) {
...
+    .questions
+    .toObject()
+    .filter(q => !q.isArchived)
```
Comment (dashersw) [inline backend/src/models/event.js]: event owners should receive archived questions — but they of course won't see them on the frontend alongside other questions, slido has a separate tab for archived questions, for example.

normal users should not receive archived questions, so this implementation should stay for normal users.

File: frontend/src/App.vue (hunk 1 above)
Comment (puskuruk) [reply frontend/src/App.vue]: Resolved 🍻

File: frontend/src/components/Ugur.vue (hunk 2 above)
Comment (puskuruk) [reply frontend/src/components/Ugur.vue]: We know this as Ugur :/ But I renamed it. 🍻

**Review by dashersw — CHANGES_REQUESTED** [review]

File: frontend/src/components/Ugur.vue (hunk 3 above)
Comment (tasarozan) [reply frontend/src/components/Ugur.vue]: 👍

File: backend/src/models/event.js (hunk 14 above)
Comment (murerkinn) [reply backend/src/models/event.js]: 👍

File: frontend/src/components/Ugur.vue (hunk 12 above)
Comment (murerkinn) [reply frontend/src/components/Ugur.vue]: 👍

File: frontend/src/components/Ugur.vue (hunk 8 above)
Comment (murerkinn) [reply frontend/src/components/Ugur.vue]: 👍

File: frontend/src/components/Ugur.vue (hunk 7 above)
Comment (murerkinn) [reply frontend/src/components/Ugur.vue]: 👍

File: frontend/src/views/EventDetail.vue (hunk 11 above)
Comment (murerkinn) [reply frontend/src/views/EventDetail.vue]: 👍

File: frontend/src/views/Monitor.vue [hunk 15]
```diff
@@ -0,0 +1,77 @@
+<template lang="pug">
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue]: Script should come first

File: frontend/src/views/Monitor.vue [hunk 16]
```diff
@@ -0,0 +1,77 @@
...
+  border-radius: 10px;
+  width: 950px;
+  display: flex;
+
+  .question {
+    flex: 10;
+
+    .question-text {
+      font-family: sans-serif;
+      font-size: 1.25rem;
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue]: This should be in pixels.

File: frontend/src/views/Monitor.vue [hunk 17]
```diff
@@ -0,0 +1,77 @@
...
+      color: aliceblue;
+    }
+
+    .question-author {
+      margin: 0;
+      margin-bottom: 0.5rem;
+      font-size: 1.1rem;
+      font-family: serif;
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue:72]: This should be a defined serif font. Can you help us @furkanilgenci?

File: frontend/src/views/Monitor.vue [hunk 18]
```diff
@@ -0,0 +1,77 @@
...
+  background-color: rgba(0, 0, 0, 0.6);
+  border-radius: 10px;
+  width: 950px;
+  display: flex;
+
+  .question {
+    flex: 10;
+
+    .question-text {
+      font-family: sans-serif;
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue:64]: You should use at least system fonts.

File: frontend/src/views/Monitor.vue [hunk 19]
```diff
@@ -0,0 +1,77 @@
...
+      font-size: 1.25rem;
+      color: aliceblue;
+    }
+
+    .question-author {
+      margin: 0;
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue]: You should write margins in one line.

File: frontend/src/views/Monitor.vue [hunk 20]
```diff
@@ -0,0 +1,77 @@
...
+}
+</script>
+
+<style lang="scss" scoped>
+.display-frame {
+  padding: 20px 30px;
+  background-color: rgba(0, 0, 0, 0.6);
+  border-radius: 10px;
+  width: 950px;
+  display: flex;
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue:58]: Why do we need to use `flex` here?

File: frontend/src/views/Monitor.vue [hunk 21]
```diff
@@ -30,6 +30,19 @@ export default {
 }
 </script>
 
+<style lang="scss">
```
Comment (puskuruk) [inline frontend/src/views/Monitor.vue:33]: This commit should be squashed into the first commit.

File: backend/src/routes/account.js [hunk 22]
```diff
@@ -5,13 +5,27 @@ const Validator = require('async-validator').default
...
 const ensureSingularity = require('../lib/ensureSingularity')
+const ensureUser = require('../lib/ensureUser')
 
 const router = express.Router()
 
 router.get('/', async (req, res) => {
   res.send(req.user)
 })
 
+router.patch('/', ensureUser, async (req, res, next) => {
```
Comment (puskuruk) [inline backend/src/routes/account.js]: This should only check if `req.user` exists.

File: backend/src/routes/account.js [hunk 23]
```diff
@@ -5,13 +5,27 @@ const Validator = require('async-validator').default
 const User = require('../models/user')
 
 const ensureSingularity = require('../lib/ensureSingularity')
+const ensureUser = require('../lib/ensureUser')
```
Comment (puskuruk) [inline backend/src/routes/account.js]: We don't need this anymore.

File: frontend/src/components/Director.vue [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+
+      const isTriggered = this.triggers
+        .split('\n')
+        .some(trigger => trigger && transcript.includes(trigger))
+
+      if (isTriggered) this.skipQuestion()
+    },
+  },
+  mounted() {
+    this.triggers = this.user.director.triggers.join('\n')
```
Comment (puskuruk) [inline frontend/src/components/Director.vue:88]: You should watch `user.director.triggers`.

File: frontend/src/components/Director.vue [hunk 25]
```diff
@@ -0,0 +1,114 @@
...
+
+    if (!speechRecognitionInstance) return
+
+    speechRecognitionInstance.removeEventListener('result', this.handleVoiceTrigger)
+    speechRecognitionInstance.removeEventListener('end', speechRecognitionInstance.start)
+
+    speechRecognitionInstance.stop()
+  },
+  watch: {
+    triggers(newTriggers, oldTriggers) {
```
Comment (puskuruk) [inline frontend/src/components/Director.vue:101]: Triggers should be saved with a button. And, if there is no diff this button needs to be disabled.

File: frontend/src/components/Director.vue [hunk 26]
```diff
@@ -0,0 +1,114 @@
...
+    ...mapState('account', ['user'])
+  },
+  props: ['pinLatestQuestion'],
+  methods: {
+    ...mapActions('event', ['archiveQuestion']),
+    ...mapActions('account', ['updateDirector']),
+    activateVoiceRecognition() {
+      const { speechRecognitionInstance } = this
+
+      speechRecognitionInstance.lang = this.user.director.language
```
Comment (puskuruk) [inline frontend/src/components/Director.vue:41]: You should check if there is user or not

Comment (puskuruk) [reply frontend/src/components/Director.vue:41]: Do we have a user check in route definition?

Comment (puskuruk) [reply frontend/src/components/Director.vue:41]: Maybe we should use `event` and `user` as props. This way we wouldn't check if the `user` exists Vue will automatically checks for us.

File: frontend/src/components/Director.vue [hunk 27]
```diff
@@ -0,0 +1,114 @@
...
+      this.archiveQuestion({
+        action: 'archive',
+        questionId: getLastPinnedQuestion()._id
+      })
+
+      if (getLastPinnedQuestion()) return
+
+      this.pinLatestQuestion()
+    },
+    handleVoiceTrigger(e) {
```
Comment (puskuruk) [inline frontend/src/components/Director.vue]: Maybe `checkAndTriggerVoiceCommands`?

File: frontend/src/components/Director.vue [hunk 28]
```diff
@@ -0,0 +1,114 @@
...
+
+      speechRecognitionInstance.addEventListener('result', this.handleVoiceTrigger)
+      speechRecognitionInstance.addEventListener('end', speechRecognitionInstance.start)
+    },
+    skipQuestion() {
+      const getLastPinnedQuestion = () => this.event.questions?.find(question => question.isPinned)
+
+      if (!getLastPinnedQuestion()) return
+
+      this.archiveQuestion({
```
Comment (puskuruk) [inline frontend/src/components/Director.vue]: Shouldn't we use `await` here? Don't we call the backend?

File: frontend/src/components/Director.vue [hunk 29]
```diff
@@ -0,0 +1,114 @@
...
+        action: 'archive',
+        questionId: getLastPinnedQuestion()._id
+      })
+
+      if (getLastPinnedQuestion()) return
+
+      this.pinLatestQuestion()
```
Comment (puskuruk) [inline frontend/src/components/Director.vue]: You should  `await` it.

File: frontend/src/views/Monitor.vue (hunk 21 above)
Comment (murerkinn) [reply frontend/src/views/Monitor.vue:33]: 👍

File: frontend/src/views/Monitor.vue (hunk 15 above)
Comment (murerkinn) [reply frontend/src/views/Monitor.vue]: 👍

File: frontend/src/views/Monitor.vue (hunk 19 above)
Comment (murerkinn) [reply frontend/src/views/Monitor.vue]: 👍

File: frontend/src/views/Monitor.vue (hunk 16 above)
Comment (murerkinn) [reply frontend/src/views/Monitor.vue]: 👍

File: backend/src/models/event.js [hunk 30]
```diff
@@ -57,15 +57,19 @@ Event.pre('save', async function (next) {
...
-      voted: q.voters.some(v => userIds.some(id => id.equals(v))),
-      ownQuestion: userIds.some(i => i.equals(q.user)),
-      user: undefined,
-      voters: undefined,
-    }
+Event.static('decorateForUser', function (event, userIds, isEventOwner) {
+  const updatedQuestions = event
+    .questions
+    .toObject()
+    .filter(q => isEventOwner ? q : !q.isArchived)
```
Comment (murerkinn) [inline backend/src/models/event.js]: We don't need isEventOwner parameter since we have userIds

File: backend/src/routes/index.js [hunk 31]
```diff
@@ -101,8 +101,9 @@ router.post('/events', ensureUser, ensureLogin, eventsValidator, async function
 router.get('/events/:eventId', ensureUser, async (req, res, next) => {
   const { id: sessionId, userId, computerId } = req.session
   const userIds = await fetchUserIdsBySingularities({ sessionId, userId, computerId })
+  const isEventOwner = req.user?._id.equals(req.event.owner)
```
Comment (murerkinn) [inline backend/src/routes/index.js]: we don't need this anymore because we will use userIds to check if user is event owner

File: backend/src/routes/index.js [hunk 32]
```diff
@@ -195,6 +196,17 @@ router.patch('/events/:eventId/questions/:questionId', ensureUser, async functio
         },
       }
       break
+
+    case 'archive':
+    case 'unarchive':
+      arrayFilters = [{ 'question._id': questionId, 'question.user': userId }]
```
Comment (murerkinn) [inline backend/src/routes/index.js]: we should remove `question.user` property

File: frontend/src/components/AskTheSpeakerForm.vue [hunk 33]
```diff
@@ -0,0 +1,35 @@
...
+      placeholder='Type your question',
+      :autoSize='{ minRows: 2, maxRows: 6 }',
+      :maxLength='280',
+      v-model='question'
+    )
+    a-input(placeholder='Your name (optional)', v-model='name', :maxLength='40')
+    a-button(type='primary', @click="$emit('sendQuestion', { name, question })" :loading='loading', icon='message') Send
+</template>
+
+<script>
```
Comment (murerkinn) [inline frontend/src/components/AskTheSpeakerForm.vue:1]: we should move script to top

File: frontend/src/components/AskTheSpeakerForm.vue [hunk 34]
```diff
@@ -0,0 +1,35 @@
+<template lang='pug'>
+  form(v-else @submit.prevent="$emit('sendQuestion', { name, question })")
```
Comment (murerkinn) [inline frontend/src/components/AskTheSpeakerForm.vue]: we should remove v-else from this line

File: frontend/src/components/AskTheSpeakerForm.vue [hunk 35]
```diff
@@ -0,0 +1,35 @@
...
+      :maxLength='280',
+      v-model='question'
+    )
+    a-input(placeholder='Your name (optional)', v-model='name', :maxLength='40')
+    a-button(type='primary', @click="$emit('sendQuestion', { name, question })" :loading='loading', icon='message') Send
+</template>
+
+<script>
+export default {
+  name: 'askTheSpeakerForm',
```
Comment (murerkinn) [inline frontend/src/components/AskTheSpeakerForm.vue]: we should name this in PascalCase

File: frontend/src/components/AskTheSpeakerForm.vue [hunk 36]
```diff
@@ -0,0 +1,35 @@
...
+
+<style lang='scss' scoped>
+form > * {
+  margin: 8px 0 !important;
+}
+
+textarea {
+  padding: 8px;
+}
+</style>
```
Comment (murerkinn) [inline frontend/src/components/AskTheSpeakerForm.vue]: we should add a newline at the end of this file

File: frontend/src/views/EventDetail.vue [hunk 37]
```diff
@@ -179,42 +185,28 @@ export default {
...
-        a-textarea(
-          placeholder='Type your question',
-          :autoSize='{ minRows: 2, maxRows: 6 }',
-          :maxLength='280',
-          v-model='question'
-        )
-        a-input(placeholder='Your name (optional)', v-model='name', :maxLength='40')
-        a-button(type='primary', @click='sendQuestion', :loading='loading', icon='message') Send
+          Director(:handlePin="pinLatestQuestion")
+      ask-the-speaker-form(v-else :loading='loading', @sendQuestion='sendQuestion($event)')
```
Comment (murerkinn) [inline frontend/src/views/EventDetail.vue]: we don't need to send loading to AskTheSpeakerForm, also we will handle sending question in AskTheSpeakerForm

File: frontend/src/views/EventDetail.vue [hunk 38]
```diff
@@ -41,9 +42,9 @@ export default {
       'pinQuestion',
       'archiveQuestion'
     ]),
-    async sendQuestion() {
+    async sendQuestion({ question, name }) {
```
Comment (murerkinn) [inline frontend/src/views/EventDetail.vue]: We need to move this function to the AskTheSpeakerForm

File: frontend/src/views/EventDetail.vue [hunk 39]
```diff
@@ -179,42 +185,28 @@ export default {
...
-              h2 Ask the speaker
-              a-textarea(
-                placeholder="Type your question"
-                :autoSize="{ minRows: 2, maxRows: 6 }"
-                :maxLength="280"
-                v-model="question"
-              )
-              a-input(placeholder="Your name (optional)" v-model="name" :maxLength="40")
-              a-button(type="primary" @click="sendQuestion" :loading="loading" icon="message") Send
+          ask-the-speaker-form(:loading='loading', @sendQuestion='sendQuestion($event)')
```
Comment (murerkinn) [inline frontend/src/views/EventDetail.vue]: we don't need to send loading and we will handle sendQuestion in AskTheSpeakerForm, so we don't need loading data in the computed properties

File: frontend/src/views/EventDetail.vue [hunk 40]
```diff
@@ -167,6 +168,11 @@ export default {
       this.previousQuestionsSortedIds = this.questions.map(q => q._id)
 
       this.questions = questions.slice()
+      this.sortQuestions()
+    },
+    showArchivedQuestions(val) {
```
Comment (murerkinn) [inline frontend/src/views/EventDetail.vue]: we should rename this as `areArchivedQuestionsShown`

File: frontend/src/views/EventDetail.vue [hunk 41]
```diff
@@ -179,42 +185,28 @@ export default {
...
           h2 Questions
             a-avatar.question-count {{ questions.length }}
           .sort-container
             div(v-if='questions.length')
-              span Sort questions by &nbsp;
-              a-radio-group(size='small', defaultValue='popular', buttonStyle='solid', :value='sortBy')
-                a-radio-button(value='popular', @click='updateSorting') Popular {{ popularSortOrderIndicator }}
-                a-radio-button(value='recent', @click='updateSorting') Recent {{ recentSortOrderIndicator }}
-                a-radio-button(value='random', @click='updateSorting') Random {{ randomSortOrderIndicator }}
+              div
```
Comment (murerkinn) [inline frontend/src/views/EventDetail.vue]: why do we need this here

File: frontend/src/components/Director.vue (hunk 26 above)
Comment (murerkinn) [reply frontend/src/components/Director.vue:41]: I've implemented a beforeEnter as a route guard in router.js (frontend)

File: backend/src/routes/account.js [hunk 42]
```diff
@@ -16,6 +14,20 @@ router.get('/',
     res.send(req.user)
   })
 
+router.patch('/', async (req, res, next) => {
+  if (!req.user) return
```
Comment (murerkinn) [inline backend/src/routes/account.js]: we should `return next({ status: 400 })`

File: backend/src/routes/account.js [hunk 43]
```diff
@@ -16,6 +14,20 @@ router.get('/',
...
+
+  const { director } = req.body
+
+  if (!director) res.sendStatus(400)
+
+  const user = await User.findByIdAndUpdate(req.session.userId, { director })
+
+  req.session.user = user
```
Comment (murerkinn) [inline backend/src/routes/account.js]: we should remove this line

File: backend/src/routes/account.js [hunk 44]
```diff
@@ -16,6 +14,20 @@ router.get('/',
...
   })
 
+router.patch('/', async (req, res, next) => {
+  if (!req.user) return
+
+  const { director } = req.body
+
+  if (!director) res.sendStatus(400)
+
+  const user = await User.findByIdAndUpdate(req.session.userId, { director })
```
Comment (murerkinn) [inline backend/src/routes/account.js]: we can use `req.user._id`instead of `req.session.userId`

File: backend/src/routes/account.js [hunk 45]
```diff
@@ -16,6 +14,20 @@ router.get('/',
...
+
+  req.session.user = user
+
+  res.status(204).send(req.user)
```
Comment (murerkinn) [inline backend/src/routes/account.js]: we can send `user` instead of `req.user`

File: backend/src/routes/index.js [hunk 46]
```diff
@@ -186,7 +186,7 @@ router.patch('/events/:eventId/questions/:questionId', ensureLogin, rateLimiter(
 
   if (action == 'like' && !req.user && computerId.startsWith('nobiri-')) return next({ status: 401 })
 
-  if (!['like', 'unlike', 'pin', 'unpin'].includes(action)) return next({ status: 400 })
+  if (!['like', 'unlike', 'pin', 'unpin', 'archive', 'unarchive'].includes(action)) return next({ status: 400 })
```
Comment (murerkinn) [inline backend/src/routes/index.js]: we need to put these in an allowedActions array

File: backend/src/routes/index.js [hunk 47]
```diff
@@ -237,7 +256,12 @@ router.patch('/events/:eventId/questions/:questionId', ensureLogin, rateLimiter(
 
   if (!event) return next(new Error('Event or question not found'))
 
-  socketServer().to(req.params.eventId).emit('questions updated')
+  if (req.user.equals(event.owner)) {
```
Comment (murerkinn) [inline backend/src/routes/index.js]: we should use a ternary here

File: frontend/src/components/AskTheSpeakerForm.vue [hunk 48]
```diff
@@ -0,0 +1,58 @@
...
+      :maxLength='280',
+      v-model='question'
+    )
+    a-input(placeholder='Your name (optional)', v-model='name', :maxLength='40')
+    a-button(type='primary', @click="sendQuestion" :loading='loading', icon='message') Send
+</template>
+
+<style lang='scss' scoped>
+form > * {
+  margin: 8px 0 !important;
```
Comment (murerkinn) [inline frontend/src/components/AskTheSpeakerForm.vue]: we should use em instead of px

File: backend/src/routes/account.js (hunk 42 above)
Comment (furkanilgenci) [reply backend/src/routes/account.js]: done

File: backend/src/routes/account.js (hunk 43 above)
Comment (furkanilgenci) [reply backend/src/routes/account.js]: done

File: backend/src/routes/account.js (hunk 44 above)
Comment (furkanilgenci) [reply backend/src/routes/account.js]: done

File: backend/src/routes/account.js (hunk 45 above)
Comment (furkanilgenci) [reply backend/src/routes/account.js]: done

File: backend/src/routes/index.js (hunk 46 above)
Comment (furkanilgenci) [reply backend/src/routes/index.js]: done

File: backend/src/routes/index.js (hunk 47 above)
Comment (furkanilgenci) [reply backend/src/routes/index.js]: done

File: frontend/src/components/AskTheSpeakerForm.vue (hunk 48 above)
Comment (furkanilgenci) [reply frontend/src/components/AskTheSpeakerForm.vue]: done

File: backend/src/routes/account.js [hunk 49]
```diff
@@ -16,6 +14,18 @@ router.get('/',
...
   })
 
+router.patch('/', async (req, res, next) => {
+  if (!req.user) return next({ status: 400 })
+
+  const { director } = req.body
+
+  if (!director) res.sendStatus(400)
+
+  const user = await User.findByIdAndUpdate(req.user._id, { director })
```
Comment (furkanilgenci) [inline backend/src/routes/account.js]: We don't need to return updated version

File: frontend/src/components/Director.vue (hunk 24 above)
Comment (furkanilgenci) [reply frontend/src/components/Director.vue:88]: done

File: frontend/src/main.js [hunk 50]
```diff
@@ -20,7 +20,9 @@ import {
   Menu,
   Form,
   Alert,
-  Empty
+  Empty,
+  Tabs,
+  Switch
```
Comment (furkanilgenci) [inline frontend/src/main.js]: Remove switch

File: frontend/src/router.js [hunk 51]
```diff
@@ -56,6 +57,15 @@ export default function init(store) {
         name: 'eventDetail',
         component: EventDetail
       },
+      {
+        path: '/events/:eventId/monitor',
+        name: 'monitor',
+        component: Monitor,
+        beforeEnter(to, from, next) {
```
Comment (furkanilgenci) [inline frontend/src/router.js]: No need for beforeEnter

File: frontend/src/store/account.js [hunk 52]
```diff
@@ -1,6 +1,7 @@
 import axios from 'axios'
 import biri from 'biri'
 
+
```
Comment (furkanilgenci) [inline frontend/src/store/account.js]: Remove this row

File: frontend/src/store/event.js [hunk 53]
```diff
@@ -31,7 +32,7 @@ const event = {
   state: () => ({
     loading: false,
     eventId: null,
-    event: {}
+    event: null
```
Comment (furkanilgenci) [inline frontend/src/store/event.js]: Should better be undefined

File: frontend/src/store/event.js [hunk 54]
```diff
@@ -31,7 +32,7 @@ const event = {
   state: () => ({
     loading: false,
     eventId: null,
```
Comment (furkanilgenci) [inline frontend/src/store/event.js]: Should better be undefined

File: frontend/src/views/EventDetail.vue [hunk 55]
```diff
@@ -226,6 +245,12 @@ export default {
     }
   }
 
+  .archive-switch {
```
Comment (furkanilgenci) [inline frontend/src/views/EventDetail.vue]: Should be removed

File: frontend/src/views/Monitor.vue (hunk 20 above)
Comment (furkanilgenci) [reply frontend/src/views/Monitor.vue:58]: bump

File: frontend/src/views/Monitor.vue (hunk 18 above)
Comment (furkanilgenci) [reply frontend/src/views/Monitor.vue:64]: Should be the correct font

File: frontend/src/views/Monitor.vue (hunk 17 above)
Comment (furkanilgenci) [reply frontend/src/views/Monitor.vue:72]: Should be the correct font

File: frontend/src/views/EventDetail.vue (hunk 55 above)
Comment (furkanilgenci) [reply frontend/src/views/EventDetail.vue]: done

File: frontend/src/main.js (hunk 50 above)
Comment (furkanilgenci) [reply frontend/src/main.js]: done

File: backend/src/routes/account.js (hunk 49 above)
Comment (furkanilgenci) [reply backend/src/routes/account.js]: done

File: frontend/src/router.js (hunk 51 above)
Comment (furkanilgenci) [reply frontend/src/router.js]: done

File: frontend/src/store/account.js (hunk 52 above)
Comment (furkanilgenci) [reply frontend/src/store/account.js]: done

File: frontend/src/store/event.js (hunk 54 above)
Comment (furkanilgenci) [reply frontend/src/store/event.js]: done

File: frontend/src/store/event.js (hunk 53 above)
Comment (furkanilgenci) [reply frontend/src/store/event.js]: done

**Review by dashersw — APPROVED** [review]
